cmd/uid_provider/client/client.go

```diff
@@ -0,0 +1,306 @@
+
+		user.Password = newPwd
+
+		if err := set(ctx, c.rdb, user); err != nil {
+			return err
+		}
+		return nil
+	}
+	return err
+}
+
+// Login receives an user and a password, and checks if they exist and match.
+// If so, it sends a request to Insprd so it can generate a new token for the
+// given user, and returns the toker if it's creation was successful
+func (c *Client) Login(ctx context.Context, uid, pwd string) (string, error) {
+	user, err := get(ctx, c.rdb, uid)
+	if err != nil {
+		return "", err
+	}
+	if pwd != string(user.Password) {
+		return "", fmt.Errorf("user and password don't match")
+	}
+
+	payload, err := encrypt(user)
+	if err != nil {
+		return "", err
+	}
+
+	token, err := requestNewToken(ctx, payload)
+	if err != nil {
+		return "", err
+	}
+
+	return token, nil
+}
+
+// RefreshToken receives a refreshToken and checks if it's valid.
+// If so, it returns a payload containing the updated user info
+// (user which is associated with the given refreshToken)
+func (c *Client) RefreshToken(ctx context.Context, refreshToken string) (auth.Payload, error) {
+	oldUser, err := decrypt(refreshToken)
+	if err != nil {
+		return auth.Payload{}, err
+	}
+
+	newUser, err := get(ctx, c.rdb, oldUser.UID)
+	if err != nil {
+		return auth.Payload{}, err
+	}
+
+	updatedPayload, err := encrypt(newUser)
+	if err != nil {
+		return auth.Payload{}, err
+	}
+
+	return updatedPayload, nil
+}
+
+// Auxiliar methods
+
+func set(ctx context.Context, rdb *redis.Client, data User) error {
+	strData, err := json.Marshal(data)
+	if err != nil {
+		return err
+	}
+
+	err = rdb.Set(ctx, data.UID, strData, 0).Err()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func get(ctx context.Context, rdb *redis.Client, key string) (User, error) {
+	var parsedValue User
+	value, err := rdb.Get(ctx, key).Result()
+
+	if err == redis.Nil {
+		return User{}, fmt.Errorf("key '%v' does not exist", key)
+	} else if err != nil {
+		return User{}, err
+	}
+
+	json.Unmarshal([]byte(value), &parsedValue)
+
+	return parsedValue, nil
+}
+
+func delete(ctx context.Context, rdb *redis.Client, key string) error {
+	numDeleted, err := rdb.Del(ctx, key).Result()
+	if err != nil {
+		return err
+	} else if numDeleted == 0 {
+		return fmt.Errorf("no items were deleted for key %v", key)
+	}
+	return nil
+}
+
+func havePermission(ctx context.Context, rdb *redis.Client, uid, pwd string) (bool, error) {
+	requestor, err := get(ctx, rdb, uid)
+	if err != nil {
+		return false, err
+	}
+	if requestor.Password != pwd {
+		return false, fmt.Errorf("invalid password for user %v", uid)
+	}
+	if requestor.Role != 1 {
+		return false, fmt.Errorf("user %v doesn't have admin permission", uid)
+	}
+	return true, nil
+}
+
+func encrypt(user User) (auth.Payload, error) {
+	refreshURL := os.Getenv("REFRESH_URL")
+	if refreshURL == "" {
+		panic("[ENV VAR] REFRESH_URL not found")
+	}
+	keyString := os.Getenv("REFRESH_KEY")
+	if keyString == "" {
+		panic("[ENV VAR] REFRESH_KEY not found")
+	}
```

Comment:
extract these and put in a struct that is initialized on starting the cluster, that way it's only checked once and there is no leaked information for the user on a request, as well as no request panics.

You can even create a struct that performs encryption and decryption that when initialized checks all the variables and stuff